fish-rust/src/common.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{
+    ffi,
+    wchar_ffi::{wstr, WCharFromFFI, WString},
+};
+use std::{ffi::c_uint, mem};
+
+/// A scoped manager to save the current value of some variable, and optionally set it to a new
+/// value. When dropped, it restores the variable to its old value.
+///
+/// This can be handy when there are multiple code paths to exit a block.
+pub struct ScopedPush<'a, T> {
+    var: &'a mut T,
+    saved_value: Option<T>,
+}
+
+impl<'a, T> ScopedPush<'a, T> {
+    pub fn new(var: &'a mut T, new_value: T) -> Self {
+        let saved_value = mem::replace(var, new_value);
+
+        Self {
+            var,
+            saved_value: Some(saved_value),
+        }
+    }
+
+    pub fn restore(&mut self) {
+        if let Some(saved_value) = self.saved_value.take() {
+            *self.var = saved_value;
+        }
+    }
+}
+
+impl<'a, T> Drop for ScopedPush<'a, T> {
+    fn drop(&mut self) {
+        self.restore()
+    }
+}
+
+pub enum EscapeStringStyle {
+    Script(EscapeFlags),
+    Url,
+    Var,
+    Regex,
+}
+
+/// Flags for the [`escape_string()`] function. These are only applicable when the escape style is
+/// [`EscapeStringStyle::Script`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct EscapeFlags {
+    /// Do not escape special fish syntax characters like the semicolon. Only escape non-printable
```

Comment:
This means that

	escape_string(value, ESCAPE_NO_PRINTABLES | ESCAPE_NO_QUOTED)

becomes

	escape_string(value, EscapeFlags{
		no_printables: true,
		no_quoted: true,
		no_tilde: false,
		symbolic: false,
	})

which is alright I guess. Maybe we want a shortcut but it's no big deal

Reply:
It might not make much of a difference, but you should be able to use struct update syntax: `EscapeFlags { no_printables: true, ..EscapeFlags::default() }`